synapse/util/ratelimitutils.py

```diff
@@ -222,20 +222,29 @@ def on_both(r: object) -> object:
             # Ensure that we've properly cleaned up.
             self.sleeping_requests.discard(request_id)
             self.ready_request_queue.pop(request_id, None)
-            wait_span_scope.__exit__(None, None, None)
-            wait_timer_cm.__exit__(None, None, None)
             return r
 
+        ret_defer.addCallbacks(on_start, on_err)
+        ret_defer.addBoth(on_both)
+
         # Tracing
         wait_span_scope = start_active_span("ratelimit wait")
         wait_span_scope.__enter__()
+
         # Metrics
         wait_timer_cm = queue_wait_timer.time()
         wait_timer_cm.__enter__()
 
-        ret_defer.addCallbacks(on_start, on_err)
-        ret_defer.addBoth(on_both)
-        return make_deferred_yieldable(ret_defer)
+        # The current logcontext is now one created by `wait_span_scope`.
+        # We must only `__exit__` it once it has been restored by `make_deferred_yieldable`.
+        def on_after_logcontext_restored(r: object) -> object:
+            wait_timer_cm.__exit__(None, None, None)
+            wait_span_scope.__exit__(None, None, None)
+            return r
+
+        d = make_deferred_yieldable(ret_defer)
+        d.addBoth(on_after_logcontext_restored)
+        return d
```

Comment:
there is possibly another way to do this:

```python
async def do_tracing():
   with start_active_span("ratelimit wait"):
       with queue_wait_timer.time():
           await make_deferred_yieldable(ret_defer)

return defer.ensureDeferred(do_tracing())
```

Reply:
That's a lot cleaner. I'll give that a try, thanks!

Reply:
Not sure it is useful, but you can also chain them:

```python
async def do_tracing():
   with start_active_span("ratelimit wait"), queue_wait_timer.time():
       await make_deferred_yieldable(ret_defer)

return defer.ensureDeferred(do_tracing())
```
I believe this is equivalent, but not 100% sure.

Reply:
I didn't know you could do that, thanks!